app/services/billing-accounts/change-address.service.js

```diff
@@ -0,0 +1,340 @@
+ *
+ * We persist the address, company and contact first because we need their IDs in order to create the new
+ * `crm_v2.invoice_account_address` record. When we create that record we also need to apply an end date to any existing
+ * invoice account addresses with a null end date. This is how the service determines which address details are current
+ * (end date is null).
+ *
+ * The object we return has all 3 entities whether they were persisted or not. This gets passed back to the UI via
+ * the controller and is what it needs to then be able to redirect the user to the correct page.
+ *
+ * @param {Date} timestamp the timestamp to be used for any date created or updated values when persisting
+ * @param {module:InvoiceAccountModel} invoiceAccount the invoice (billing) account having its address changed
+ * @param {module:AddressModel} address the new address to be persisted (expected to be populated)
+ * @param {module:CompanyModel} company the new agent company to be persisted (not expected to be populated)
+ * @param {module:ContactModel} contact the new contact to be persisted (not expected to be populated)
+ *
+ * @returns {Object} a single object that contains the persisted invoiceAccountAddress, plus address, agent company and
+ * contact
+ */
+async function _persist (timestamp, invoiceAccount, address, company, contact) {
+  const persistedData = {}
+
+  await InvoiceAccountModel.transaction(async (trx) => {
+    persistedData.address = await _persistAddress(trx, address)
+    persistedData.agentCompany = await _persistCompany(trx, company)
+    persistedData.contact = await _persistContact(trx, contact)
+
+    const invoiceAccountAddress = InvoiceAccountAddressModel.fromJson({
+      invoiceAccountId: invoiceAccount.invoiceAccountId,
+      addressId: persistedData.address.addressId,
+      agentCompanyId: persistedData.agentCompany.companyId,
+      contactId: persistedData.contact.contactId,
+      startDate: timestamp,
+      endDate: null,
+      createdAt: timestamp,
+      updatedAt: timestamp
+    })
+
+    await _patchExistingInvoiceAccountAddressEndDate(trx, invoiceAccount.invoiceAccountId, timestamp)
+    persistedData.invoiceAccountAddress = await _persistInvoiceAccountAddress(trx, invoiceAccountAddress)
+  })
+
+  return persistedData
+}
+
+async function _patchExistingInvoiceAccountAddressEndDate (trx, invoiceAccountId, timestamp) {
+  // The timestamp represents the current date and time we're making this change, i.e. today. So, the new invoice
+  // account address will start from today. To show that the old record is no longer current, we need to set its
+  // `endDate` to be today - 1 (yesterday). The following works it all out even if we're over a month or year boundary
+  // and no moment() in sight! Thanks to https://stackoverflow.com/a/1296374 for how to do this
+  const endDate = new Date()
+  endDate.setDate(timestamp.getDate() - 1)
+
+  return InvoiceAccountAddressModel.query(trx)
```

Comment:
Do we need to use `return` here seeing as we don't use the return value?
```suggestion
  await InvoiceAccountAddressModel.query(trx)
```

Reply:
Agreed, but think I need to have another convo with you, the team Node maestro 🤓, about what we should be doing in these cases.